### PR TITLE
feat(tests): generate folding-ranges doc from snapshot fixtures

### PR DESCRIPTION
**Pull request by 16bit-ykiko**
## Why

Feature docs are hand-maintained support checklists with no mechanical connection to tests: nothing verifies that a checked item actually works, and nothing flags when behavior changes. This PR pilots the inversion for folding ranges: snapshot fixtures become the single source of truth and the doc checklist is generated from them.

## Design

Each fixture under `tests/data/folding_range/<section>/` carries a doc header that is a plain markdown document behind `///`:

```cpp
/// # Preprocessor conditional folding (`#if` / `#ifdef` ...)
///
/// - status: partial
/// - issues: clangd#1661, clangd#2059
///
/// Branch regions delimited by `#else` fold today; a bare
/// `#if ... #endif` block without an `#else` does not fold yet.

<example code — doubles as the doc example>
```

- The h1 title marks the file as a doc item (files without it are supplementary tests, excluded from docs); stripped of the `///` prefix the whole header renders as-is in any markdown viewer.
- Grouping comes from one-level section subdirectories (`fold_kinds/`, `refinements/`); generated doc regions are keyed by the directory name, while human headings and prose stay hand-written.
- `status` semantics: `supported` → `[x]`, compiled, snapshot proves the behavior; `partial` → `[ ]` with a _(partial)_ marker, compiled, snapshot records the current partial behavior so improvements surface as diffs; `unsupported` → skipped by the snapshot glob, snapshot pinned to `UNSUPPORTED`, implementing the capability later surfaces as a diff prompting the status flip.
- `tests/tools/feature_docs.py` renders headers into marker-delimited regions of `docs/en/features/folding-ranges.md` (`update`) and fails on drift (`check`, now part of `pixi run unit-test`). The C++ side reads only the `status` key via a shared ten-line helper — there is no grammar to drift between the two parsers.

## Notable corrections found by review

- Two preprocessor fixtures initially marked unsupported actually fold `#else`-delimited branches today; they are now `partial` with snapshots recording the true current behavior (bare `#if...#endif` still does not fold — clangd#1661). The `partial` tier exists because of this finding.
- All supported items' snapshots were verified to demonstrate their claimed capability.

## Migration

The 14 checklist items of the old hand-written doc are preserved (titles, statuses, examples, issue links, client-support notes). Legacy inline test cases are untouched and migrate in follow-up per-feature PRs.

## Tests

- Unit: 1052 passed on both RelWithDebInfo and Debug, including the doc-sync check
- Integration: 300 passed; Smoke: 3/3
- `format` → `check` round-trip verified stable

### CODE REVIEW

File: tests/unit/feature/folding_range_tests.cpp [hunk 1]
```diff
@@ -437,6 +437,28 @@ TEST_CASE(snapshot) {
...
+                while(!rest.empty()) {
+                    auto parts = rest.split('\n');
+                    rest = parts.second;
+                    auto line = parts.first.trim();
+                    if(!line.starts_with("///")) {
+                        break;
+                    }
+                    line = line.drop_front(3).trim();
+                    if(line.consume_front("@status") && line.trim() == "unsupported") {
+                        return "UNSUPPORTED";
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/unit/feature/folding_range_tests.cpp]: Deliberate trade-off: `unsupported` fixtures are skipped by design (zest's glob has no per-file skip, and compiling them would conflate "capability missing" with "input produces unrelated output" — exactly the confusion we hit with the preprocessor fixtures, now modeled as `partial`). The status flip is owned by whoever implements the capability: the generated checklist points at the fixture, and the `partial` tier exists for anything observably in between. If zest grows per-file skip reporting we can revisit.

File: tests/data/folding_range/inactive_preprocessor_branch.cpp [hunk 2]
```diff
@@ -0,0 +1,19 @@
+/// @section Refinements
+/// @title Inactive preprocessor branch indication — visually distinguish or auto-fold inactive `#if`/`#else` branches
+/// @status partial
+/// @order 3
+///
+/// The server already emits fold ranges for both branches of an
+/// `#if`/`#else`, so clients can fold either branch manually. Knowing which
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/data/folding_range/inactive_preprocessor_branch.cpp]: Fixed in faece4ea — the description now states that only the condition-to-`#else` region folds and the post-`#else` branch gets no range yet.

File: tests/data/folding_range/comment_folding.cpp [hunk 3]
```diff
@@ -0,0 +1,13 @@
+/// @section Fold Kinds
+/// @title Comment folding — multi-line `/* */` and consecutive `//` line comments
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/data/folding_range/comment_folding.cpp]: Good catch — fixed in faece4ea. The snapshot transform now strips the leading `///` spec header before compiling, so header comments can never contribute ranges, and snapshot positions align with the example code shown in the generated doc. This also fixed a Debug-only CI abort: the test framework's annotation parser asserted on `@section`/`@title` tokens in the header.

File: tests/data/folding_range/fold_from_declaration_line.cpp [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+/// > **Client support**: this depends on the client interpreting
+/// > `FoldingRange.startLine` correctly. VS Code uses the line _after_
+/// > `startLine` as the first hidden line, so setting `startLine` to the
+/// > declaration line achieves the desired effect. However, VS Code still
+/// > leaves the closing `}` on a separate line rather than collapsing it onto
+/// > the signature line ([vscode#3352](https://github.com/microsoft/vscode/issues/3352)
+/// > — still open). Other clients may differ.
+
+// folded: void processData(const Config& cfg) {...}
+// not:    {... (signature hidden above fold)}
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/data/folding_range/fold_from_declaration_line.cpp]: Fixed in faece4ea — the fixture now contains a real multi-line struct and function definition, so flipping it to supported will require a snapshot whose start positions actually prove declaration-line folding.

File: tests/unit/feature/folding_range_tests.cpp [hunk 5]
```diff
@@ -437,7 +437,32 @@ TEST_CASE(snapshot) {
...
+            // are pinned to an explicit UNSUPPORTED marker instead of
+            // being compiled: zest's glob has no per-file skip.
+            auto buffer = llvm::MemoryBuffer::getFile(path);
+            if(!buffer)
+                return "COMPILE_ERROR";
+
+            llvm::StringRef source = (*buffer)->getBuffer();
+            while(source.ltrim().starts_with("///")) {
+                auto parts = source.split('\n');
+                source = parts.second;
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/unit/feature/folding_range_tests.cpp]: Fixed — the strip now only applies when the leading `///` block carries at least one `@key` line, mirroring parse_fixture's supplementary-fixture rule; a plain doc comment stays in the compiled input.

File: tests/unit/feature/folding_range_tests.cpp [hunk 6]
```diff
@@ -437,7 +437,40 @@ TEST_CASE(snapshot) {
...
+
+            llvm::StringRef source = (*buffer)->getBuffer();
+            llvm::StringRef rest = source;
+            bool spec_header = false;
+            while(rest.ltrim().starts_with("///")) {
+                auto parts = rest.split('\n');
+                rest = parts.second;
+                auto line = parts.first.trim().drop_front(3).trim();
+                if(line.starts_with("@")) {
+                    spec_header = true;
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/unit/feature/folding_range_tests.cpp]: Fixed — the C++ detector now mirrors parse_fixture's grammar: once a blank or prose line ends the key block, later `@` tags (e.g. Doxygen `@param`) no longer mark the block as a spec header, so such supplementary fixtures keep their leading comments in the compiled input.

File: tests/data/folding_range/fold_kinds/include_region.cpp [hunk 7]
```diff
@@ -0,0 +1,11 @@
+/// @section Fold Kinds
+/// @title Include region folding — consecutive `#include` directives
+/// @status unsupported
+/// @order 8
+
+#include <vector>       // ┐
+#include <string>       // │ foldable region
+#include <algorithm>    // ┘
+
+#include "app.h"        // ┐ separate region
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/data/folding_range/fold_kinds/include_region.cpp:10]: Acknowledged, deliberately deferred: this fixture is skipped today, and flipping it to supported/partial without resolvable inputs yields a loud COMPILE_ERROR snapshot — it cannot be silently approved as green. Making multi-file inputs (sibling headers in VFS) expressible is the planned multi-file fixture format, which include-region folding will need anyway; keeping the canonical doc example here beats contorting it to compile in the single-file VFS path now.

File: tests/tools/feature_docs.py [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+        return None
+
+    for key in REQUIRED_KEYS:
+        if key not in keys:
+            problems.append(f"{path}: missing required @{key}")
+    for key in keys:
+        if key not in KNOWN_KEYS:
+            problems.append(f"{path}: unknown key @{key}")
+    status = keys.get("status", "")
+    if status and status not in VALID_STATUS:
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/tools/feature_docs.py]: Fixed — @status is now validated whenever the key is present, so a bare `/// @status` is reported as invalid instead of silently rendering as unchecked.

File: tests/tools/feature_docs.py [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+    in_desc = False
+    for raw in header:
+        content = strip_comment(raw)
+        if not in_desc:
+            if content.strip() == "":
+                in_desc = True
+                continue
+            match = re.match(r"@(\w+)\s*(.*)", content)
+            if match:
+                keys[match.group(1)] = match.group(2).strip()
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/tools/feature_docs.py]: Fixed — duplicate keys are now reported as problems instead of last-one-wins, so a contradictory double @status fails check.

File: tests/unit/feature/folding_range_tests.cpp [hunk 10]
```diff
@@ -437,7 +437,49 @@ TEST_CASE(snapshot) {
...
+            // zest's glob has no per-file skip.
+            auto buffer = llvm::MemoryBuffer::getFile(path);
+            if(!buffer)
+                return "COMPILE_ERROR";
+
+            llvm::StringRef source = (*buffer)->getBuffer();
+            llvm::StringRef rest = source;
+            bool spec_header = false;
+            bool in_description = false;
+            while(rest.ltrim().starts_with("///")) {
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/unit/feature/folding_range_tests.cpp]: Fixed — the scan now peeks at each physical line and stops at the first non-/// line (including blanks), so /// doc comments in the body stay compiled input and the empty-line drop_front assert is gone. Verified with a Debug-build probe fixture.

File: tests/tools/feature_docs.py [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+            in_desc = True
+        desc.append(content)
+
+    if not keys:
+        # No spec keys at all: supplementary edge-case fixture, not a doc item.
+        return None
+
+    for key in REQUIRED_KEYS:
+        if key not in keys:
+            problems.append(f"{path}: missing required @{key}")
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/tools/feature_docs.py]: Fixed — all required keys are now rejected when present but empty, so a bare @title can no longer publish a blank checklist entry.

File: tests/unit/feature/folding_range_tests.cpp [hunk 12]
```diff
@@ -437,7 +437,56 @@ TEST_CASE(snapshot) {
...
+                // in the body after the separator stay compiled input.
+                if(!line.starts_with("///")) {
+                    break;
+                }
+                rest = parts.second;
+                line = line.drop_front(3).trim();
+                if(in_description) {
+                    continue;
+                }
+                if(!line.starts_with("@")) {
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/unit/feature/folding_range_tests.cpp]: Fixed — a key line now requires a word character immediately after `@`, mirroring parse_fixture's `@(\w+)` grammar, so `/// @ note` style prose keeps the file supplementary on both sides.

File: tests/tools/feature_docs.py [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+
+    out = [line]
+    if fx.description:
+        out.append("")
+        out.extend(indent(fx.description))
+    if fx.example:
+        out.append("")
+        out.append("  ```cpp")
+        out.extend(indent(fx.example))
+        out.append("  ```")
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/tools/feature_docs.py]: Fixed — the generator now sizes the fence to exceed the longest backtick run in the example, so example content can never terminate the fence early. Current fixtures render identically.

File: tests/tools/feature_docs.py [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+        match = BEGIN_RE.search(line)
+        if not match:
+            out.append(line)
+            idx += 1
+            continue
+
+        section = match.group(1).strip()
+        doc_sections.add(section)
+        end = idx + 1
+        while end < len(lines) and END_MARKER not in lines[end]:
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/tools/feature_docs.py]: Fixed — both BEGIN and END markers now match only as standalone unindented lines, so marker text embedded in generated item content (which is always indented or inline) can never open or terminate a region.

File: tests/tools/feature_docs.py [hunk 15]
```diff
@@ -0,0 +1,374 @@
...
+    while idx < len(lines):
+        line = lines[idx]
+        match = BEGIN_RE.search(line)
+        if not match:
+            out.append(line)
+            idx += 1
+            continue
+
+        section = match.group(1).strip()
+        doc_sections.add(section)
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/tools/feature_docs.py:295]: Fixed — a second region marker for the same section is now reported as a problem instead of silently rendering the checklist twice.

File: tests/tools/feature_docs.py [hunk 16]
```diff
@@ -0,0 +1,374 @@
...
+    while i < len(lines) and lines[i].lstrip().startswith("///"):
+        desc.append(strip_comment(lines[i]))
+        i += 1
+    body_start = i
+
+    # Doc-item fixtures compile verbatim and the annotation parser swallows
+    # `$` as a nameless point marker, so the doc example would silently
+    # diverge from the compiled input. Bare `@` is fine (Doxygen tags);
+    # only the `@key[...]` shape is an annotation.
+    if "$" in text:
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/tools/feature_docs.py]: Resolved at the root by the annotation-syntax redesign in #537: the test annotation grammar no longer reserves `@` (or any ASCII character), so Doxygen tags in fixtures pass through the parser untouched in every build mode. Once #537 lands, this branch will be rebased and the `$` reserved-character check updated to the new sigils.

File: tests/unit/feature/folding_range_tests.cpp [hunk 17]
```diff
@@ -437,6 +438,16 @@ TEST_CASE(snapshot) {
...
+            // UNSUPPORTED marker instead of being compiled: zest's glob
+            // has no per-file skip. Everything else compiles verbatim —
+            // the frontmatter is an ordinary comment to the compiler.
+            auto buffer = llvm::MemoryBuffer::getFile(path);
+            if(!buffer)
+                return "COMPILE_ERROR";
+            if(fixture_frontmatter((*buffer)->getBuffer(), "status") == "unsupported")
+                return "UNSUPPORTED";
+
             if(!compile_file(path))
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/unit/feature/folding_range_tests.cpp:451]: Deliberate: the key-block frontmatter compiles as an ordinary comment by design — that is what removed the C++-side header parser and its grammar-sync problems. When comment folding is implemented, the corpus regenerates once and the header folds are honest parts of the input; fixtures that want to exclude them can use the <snap:begin>/<snap:end> region filter introduced in #537.

File: tests/unit/test/fixture.h [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+///     ///
+///     /// Optional markdown description after a bare `///` separator.
+///
+/// A file has a header iff its first line (trimmed) starts with `/// #`.
+/// Returns the value of the metadata list entry `- key: value` inside the
+/// leading run of `///` lines, or an empty StringRef when the file has no
+/// header or the key is absent. This is a deliberately trivial scan; the
+/// Python parser is the authority on the full grammar.
+inline llvm::StringRef fixture_frontmatter(llvm::StringRef content, llvm::StringRef key) {
+    if(!content.ltrim().starts_with("/// #")) {
```
[the remark this answers is not in the document]
Comment (16bit-ykiko) [reply tests/unit/test/fixture.h]: Fixed — the C++ helper now applies parse_fixture's exact title grammar (strip /// plus one space, require `# `), so both sides classify these variants identically.
